### PR TITLE
http-data: store only results of fromApi, fire change event only once 

### PR DESCRIPTION
**Pull request by jsnajdr**
Before this patch, the `onSuccess` handler in `http-data` saved the raw response under the request ID key, and then looped through the array returned from `fromApi` to store additional IDs.

After this patch, we store only results from `fromApi`. The default implementation simply returns the raw response, so in this case the result was previously stored twice.

With non-default `fromApi` implementation, it could happen that two results (raw and from `fromApi`) could be stored in quick succession under the same ID.

The second thing this patch does is to call the listeners only after all updates are done, and not after each update separately.

This change modifies the behavior of `waitForHttpData`, which was previously somewhat counter-intuitive: storing the raw response under `requestId` immediately fired the change event and resolved the `waitForHttpData` promise, so the consumer got the raw response. But immediately after this, the non-raw response from `fromApi` was stored under the same ID. The original `waitForHttpData` never saw this.

But if you called the same `waitForHttpData` again, it would return a different (cached) response, not the same one as the first call. That's a bug that wasn't happening in production, but was still present and dangerous.

The fact that we trigger the listeners only once will help us build a `useHttpData` hook that will remove a big chunk of Redux from the `http-data` library -- no more `HTTP_DATA_TICK` action, as the hook will subscribe a `http-data` listener.

There are two more commits that clean up the usages of `waitForHttpData` and align the expected shapes of data.

**How to test:**
- `waitForHttpData` in Reader: go to `/read/feeds/site.slug` and verify that you're redirected to a numeric feed ID URL
- `waitForHttpData` in Signup: I tested this by commenting out various conditions in the `redirectTests` controller and verified with debugger that the geolocation returned is correct when handing a `/start/*` route.

### CODE REVIEW

**Comment by dmsnell** [comment]
> The fact that we trigger the listeners only once will help us build a useHttpData hook that will remove a big chunk of Redux from the http-data library -- no more HTTP_DATA_TICK action, as the hook will subscribe a http-data listener.

🎉 this is amazing. at one point I also had a patch for this that used a `Context.Provider` which might have some merit, though it does seem like a better fit for hooks given that it updates on its own.

> saved the raw response under the request ID key, and then looped through the array returned from fromApi to store additional IDs.

while I can't say this is the ideal situation it wasn't an oversight. you might want to double-check that nothing still depends on this behavior. it is strange that it quickly overwrites the request id but at the same time I don't think we have to assume that the id given in the request will match the ids returned by `fromApi()`, especially with things like `POST` requests or when fetching lists of things.

my goal was to have a way of keying the request separately from the resources and being able to know the status of that request.

```ts
const sitePostsRequest = requestHttpData(`site-posts-${siteId}`, …);
// we can display that we are fetching the posts or _have_ fetched them already
```

then with this we get a list of pairs of the posts keyed individually which post-rendering components can use instead of depending on the full list

```ts
const postRequest = requestHttpData(`site-post-${siteId}-${postId}`, …);
// fulfilled by the parent component already with the post-list request
// but if we go here directly or the `freshness` demands we'll fetch just this post
```

I guess I care less about the raw response (though it's helpful at times) and more about keeping separate the request meta and lifecycle of the id from the `requestHttpData()` call and that from the returned data points. in most cases I expect people to return an id from `fromApi()` that matches the one from `requestHttpData()` but I'd hope there'd be enough use where this isn't the case that it's normative.

---

I hope this context is helpful. it does seem right in any case to prevent a double update in quick succession if we know we'll get  one, and making Redux an add-on is right in line with the vision for this as a light-weight declarative API utility - something only not done because we were in a time crunch when this was written. thanks as always!

**Comment by jsnajdr** [comment]
> you might want to double-check that nothing still depends on this behavior.

I checked that earlier when doing other fixes for `http-data` in #46193, #46235 or #46305. These fixed a wrong default implemenation of `fromApi`, and also fixed some usages where the consumers supplied `fromApi` functions that were obviously wrong.

But I plan to check everything once again before merging this PR.

> my goal was to have a way of keying the request separately from the resources and being able to know the status of that request.

Fun fact: there were several `fromApi` implementations that tried to do this, but I don't remember seeing any that did it correctly 🙁

After this patch, there are two broad ways how to use `requestHttpData`:

1. With the default `fromApi`, the REST response is stored under the request key. There is 1:1 correspondence, no magic.

2. With custom `fromApi`, we can make things more independent from each other:

Request site list, fetch site list, store list and individual items:
```js
requestHttpData( 'sites', http( '/sites' ), { fromApi : function*( response ) {
  yield [ 'sites', response.sites ];
  for ( const site of response.sites ) {
    yield [ 'site-' + site.ID, site ];
  }
} } );
```
(I use generator as `fromApi`, because the code is then nicer, avoiding array spreads)

Request individual site, fetch the whole site list, store as in previous example:
```js
requestHttpData( 'site-1', http( '/sites' ), { fromApi: sameAsAbove } );
```

Request individual site, fetch individual site. Uses cached version from previous requests for the whole list, simple enough to use default `fromApi`:
```js
requestHttpData( 'site-1', http( '/site/1' ) );
```

That looks good, doesn't it? Note that the example with `requestHttpData( 'site-1', http( '/sites' ) )` works reliably only with this PR applied. Before that, the whole sites list would be stored into `site-1`, only to be updated a bit latter to the correct individual site object.

**Review by tyxla — APPROVED** [review]
I've done some testing and everything affected appears to be working great.

Really nice cleanup! From a functionality standpoint, it might turn out to be useful to have access to more than just the response body, but if we don't need it right now... there's a chance we might never need it 😉  I generally like getting rid of the extra code we don't need in Calypso, so this looks really cool 👍 

That being said, I'm happy to defer to Dennis for the final decision here, as he's the one who put most of the pre-existing work in.

**Comment by jsnajdr** [comment]
> From a functionality standpoint, it might turn out to be useful to have access to more than just the response body

I don't understand -- what else would you want to access other than the entire response body, which is saved by the default `fromApi`? The response headers, for example?

**Comment by tyxla** [comment]
> The response headers, for example?

Yes, I had exactly the headers in mind.

**Comment by dmsnell** [comment]
> But I plan to check everything once again before merging this PR.

Thanks. I don't think you'll find any either but I felt like at some point something _was_ using it.

> (I use generator as fromApi, because the code is then nicer, avoiding array spreads)
 
Clever use of generators without changing the system specifically for them. I hope we don't end up using them or documenting them as such. I see that you're using a pretty benign case to iterate through a list synchronously but I don't even know how to think through what would happen if people start using timeouts or network fetches simply because they can.

It's an array spread or a `for` loop.

> That looks good, doesn't it? Note that the example with requestHttpData( 'site-1', http( '/sites' ) ) works reliably only with this PR applied. Before that, the whole sites list would be stored into site-1, only to be updated a bit latter to the correct individual site object.

I'm lost following this: the way it's written _intends_ on storing all the sites in `site-1` and then overwriting it with the individual site data. Are you saying that the danger are bugs like calling `requestHttpData('site-1', http('/sites'))` when that should have only been `requestHttpData('site-1', http('/sites/1'))`?

---

As far as storing the data in the request id and updating twice wouldn't it be harmless to leave that in here since it would plausibly be overwritten before the listeners are called? Or is the problem that you overwrite the request id without realizing it when you specify a list of ids in `fromApi()` that aren't the request id?

**Comment by jsnajdr** [comment]
> I'm lost following this: the way it's written _intends_ on storing all the sites in `site-1` and then overwriting it with the individual site data.

When writing:
```js
const siteState = requestHttpData( 'site-1', http( '/sites' ), fromApi )
```
I understand it as follows:
- I want the function to return data for `site-1` because that's what I want to work with. The first argument, first and foremost, says _what data the caller wants_.
- the remaining arguments are a prescription how to get the data. The `http( '/sites' )` arg says that we'll be fetching the whole list.
- finally, `fromApi` is the function that controls mapping the response to the actual cache keys like `sites` or `site-1`.

Given the above, the _intent_ is to return data for `site-1`, and the intent to store stuff is encoded inside `fromApi`.

**Comment by jsnajdr** [comment]
> Are you saying that the danger are bugs like calling `requestHttpData('site-1', http('/sites'))`

It's OK to fetch an entire list and then return only one item. The point is that if we already have a cached value for 'site-1', then we will return it without issuing a request. And the value could come from both `http('/sites')` and `http('/site/1')`. I think flexibility like this is very desired with list-based REST APIs.

> As far as storing the data in the request id and updating twice wouldn't it be harmless to leave that in here since it would plausibly be overwritten before the listeners are called?

This approach already caused one bug: `waitForHttpData` finishing after the first update and never waiting for the second one.

Also, the double update is controversial once you start assigning types to things. What is the type of the `site-1` value in the cache? It should be `Resource<Site>`. But our code would first assign a `Resource<SiteList>` value, and only the second update would assign value of the correct type. The code that does two updates is not typesafe.

**Comment by dmsnell** [comment]
> The http( '/sites' ) arg says that we'll be fetching the whole list.

This is backwards for this API, because `https('/sites')` is a bug. It's right that the intent is to get `site-1` but we must avoid bloating and confusing the API usage by making indirect or implicit calls. This is a REST-focused API and so it's not entirely right to say `the remaining arguments are a prescription how to get the data` - the remaining arguments specify where the resource is located and how to transform its value into Calypso. The resource might contain other named resources and it's okay to store/cache those on success.

Really, if we say that `requestHttpData('site-1', http('/sites'), …)` is a legitmate way to call this then we have broken down one of the few motivating reasons this was designed for in the first place. It's not the `site-1` resource at `/sites` but the `site-list` (or whatever its name is). This may not seem like a big deal _because the code will still "work" if you do it wrong_ but if we ever do start using this more broadly, which is a great way to break apart state and remove a lot of needless code, then we will start fighting each other when we say we are getting one thing but actually getting another. Besides, `site-1` may not even be in `/sites` - and why must it? If we wanted `site-1` we should have asked for `site-1`.

**Comment by jsnajdr** [comment]
@dmsnell I see one reason why `requestHttpData( 'site-1', https( '/sites' ) )` could have unexpected results: the site with ID 1 is not necessarily part of the sites list, so the state of the `site-1` will not be set. It will remain "uninitialized" instead of the expected "error: not found".

Other than that, does there need to be an exact 1:1 mapping between the requested resources IDs and REST API paths? The `/sites` endpoint returns not only the `sites-list` resource, but also all the individual `site-{id}` ones.

I can imagine situations where, for example, an older Jetpack version doesn't have the `/site/:id` endpoint and we need to fetch the whole list to get an individual site.

That's suboptimal for sure, but I don't see how it could ever break the entire system or anything like that. The `http( '/sites' )` does a legitimate best effort to get `site-1` to us, so why not?

**Comment by dmsnell** [comment]
> Other than that, does there need to be an exact 1:1 mapping between the requested resources IDs and REST API paths?

This is a bit like asking if there needs to be a mapping between a REST resource and its location. Of course not, but the system doesn't exist to make it easy to make HTTP calls - it exists to make it easy to request resources of a given name. We generally don't hit the posts-list endpoint just to get a post even though that too would work.

Just one superficial example - someone carelessly grabs the site list and calls it `site-1` because that seemed to work when they tried it but then someone else wanted `sites-list` and duplicated the fetch needlessly because the resource and its location were split apart. If the original developer had _just_ requested `site-1` then they would have gotten a smaller and faster API response with just the data they wanted and the second call wouldn't have been a duplicate since it needed more information. Alternatively if that original developer had been honest about the request and grabbed `site-1` from the `sites-list` resource then the second call would have been avoided entirely, having already been fulfilled by the local cache.

> I can imagine situations where, for example, an older Jetpack version doesn't have the /site/:id endpoint and we need to fetch the whole list to get an individual site.

This isn't any different with `httpData` than any other code. Even if this were the case we'd still expect people to make it clear in their code why they are grabbing an entire site list just to get a single site. `httpData()` hides the location of the resource in order to present its interface based solely on the resource name. It's not going to be obvious to anyone else that when you call `requestSite(14)` that you're actually requesting everything.

We're not talking about whether this will break things. It's going to work. Does it make the system better or help people understand the intended use of the system? Does it lead to confusing inefficiencies? It's one thing if the endpoint simply doesn't exist - we can cover that with comments and explanations for the abnormality. It's another thing to advertise a misleading use of the system when we don't have that need. This is a REST-focused system: `[Resource Name, Where to Find it, How to transform it into Calypso]`

**Proper code with the mismatch between resource and location**

This IMO is how we should handle cases where we have to go so some list in order to get an individual item. It still gets the whole site list but preserves the notion that the sites-list and individual site resources are separate things.

```ts
const requestSites = () => requestHttpData(
	'sites',
	http('/sites'),
	(data) => [
		['sites', data.sites.map(site => site.id)],
		...data.sites.map(site => [`site-${site.id}`, site])
	]
);

const requestSite = (siteId) => {
	requestSites();
	return getHttpData(`site-${siteId}`);
}
```

**Pull request merged**
